config/sql/migration/migrate.25.sql

```diff
@@ -0,0 +1,9 @@
+-- Don't modify this! Create a new migration instead--see docs/ota-client-guide/modules/ROOT/pages/schema-migrations.adoc
+SAVEPOINT MIGRATION;
+
+CREATE TABLE device_data(unique_mark INTEGER PRIMARY KEY CHECK (unique_mark = 0), hardware_info_hash TEXT NOT NULL DEFAULT "", installed_packages_hash TEXT NOT NULL DEFAULT "", network_info_hash TEXT NOT NULL DEFAULT "", configuration_hash TEXT NOT NULL DEFAULT "");
```

Comment:
I'm starting to dislike this kind of one-row tables.

Maybe we can actually simplify things here with something like a key/value:

```
CREATE TABLE device_data(data_type TEXT PRIMARY KEY, hash TEXT NOT NULL);
```

And then, we don't even have to duplicate the load/store methods completely, they can call these with data_type as a constant:

```
void storeDeviceDataHash(const std::string &data_type, const std::string &hash);
bool loadDeviceDataHash(const std::string &data_type, std::string *hash);
```

Plus, each of these can just use `INSERT OR REPLACE` instead of checking if the table is empty first.

Reply:
Yeah, I was debating whether I should add these new items to the existing `device_info` table. I decided against it just because it seemed like two separate concepts. This seems like it would be more flexible (as in, fewer migrations necessary to add new items). Should we migrate some of the other similar things to that model as well?

Reply:
That's a fair question. Perhaps I was too vague about what felt wrong but it's probably that these hashes are pretty independent from each others and read/written at different times. And they have the same format, so it felt more natural to have them in different rows.

As far as other examples we have, the primary_keys table looks similar but I think the public/private keys are not so independent and we store both at once. It also looks like the other "one-row" tables are updated all at once, with fields of various types.

Also (again, opinion :D), the pain threshold we should have before changing the table schema should be slightly higher than changing regular code as they are not completely free to migrate (resources as much as risk). So I don't think we have anything urgent to change there right now.